pkg/healthchecker/health_checker.go

```diff
@@ -168,3 +162,61 @@ func execCommand(timeout time.Duration, command string, args ...string) (string,
 	}
 	return strings.TrimSuffix(string(out), "\n"), nil
 }
+
+// kubeletHttpHealthCheck checks the health api response on kubelet.
+// Returns true for healthy, false otherwise.
+func kubeletHttpHealthCheck(healthCheckTimeout time.Duration) bool {
+	httpClient := http.Client{Timeout: healthCheckTimeout}
+	response, err := httpClient.Get(types.KubeletHealthCheckEndpoint)
+	if err != nil || response.StatusCode != http.StatusOK {
+		return false
+	}
+	return true
+}
+
+// kubeletConnectionHealthCheck checks for the kubelet-apiserver connection issue
+// by checking repeated occurrences of log "use of closed network connection" in kubelet logs.
+// Returns true if the pattern does not exist 10 times since start of service or the last 10 min, false otherwise.
+func kubeletConnectionHealthCheck() (bool, error) {
+	kubeletUptimeFunc := getUptimeFunc(types.KubeletComponent)
+	uptime, err := kubeletUptimeFunc()
+	if err != nil {
+		return true, err
+	}
+	logStartTime := time.Now().Add(-uptime).Format(types.LogParsingTimeLayout)
+	if err != nil {
+		return true, err
+	}
+	out, err := execCommand(types.CmdTimeout, "/bin/sh", "-c",
+		// Query kubelet logs since the logStartTime
+		"journalctl --unit kubelet --since \""+logStartTime+"\" "+
+			// Grep the pattern for lost connection
+			"| grep -i \""+types.KubeletClosedConnectionLogPattern+"\" "+
+			// Get the count of occurrences
+			"| wc -l")
+	if err != nil {
+		return true, err
+	}
+	occurrences, err := strconv.Atoi(out)
+	if err != nil {
+		return true, err
+	}
+	glog.Infof("log occurrence pattern count : %v", occurrences)
+	return occurrences < 10, nil
+}
+
+// getKubeletHealthCheckFunc returns a function that checks for kubelet health and
+// return false if identified as unhealthy, true otherwise.
+func getKubeletHealthCheckFunc(healthCheckTimeout time.Duration) func() bool {
+	return func() bool {
+		httpHealthy := kubeletHttpHealthCheck(healthCheckTimeout)
+		// If there is any error while checking for health, log and return healthy
+		// because we do not know if the problem is in kubelet or not.
+		connectionHealthy, err := kubeletConnectionHealthCheck()
+		if err != nil {
+			glog.Infof("Error in determining apiserver connection health: %v", err)
```

Comment:
The plugin should return `unknown` when the health detection failed unexpected.

Reply:
Makes sense. Propagated the error up and eventually the plugin returns unknown when error occurs.